protobuf-maven-plugin/src/main/java/io/github/ascopes/protobufmavenplugin/sources/incremental/IncrementalCacheManager.java

```diff
@@ -0,0 +1,212 @@
+
+    // If any sources were deleted, we should rebuild everything, as those files being deleted may
+    // have caused a compilation failure.
+    if (filesDeletedSinceLastBuild) {
+      log.info("Detected that source files have been deleted, all sources will be recompiled");
+      return flattenSourceProtoFiles(listing.getCompilableSources());
+    }
+
+    return newBuildCache.getSources().keySet()
+        .stream()
+        .filter(file -> !Objects.equals(
+            newBuildCache.getSources().get(file),
+            previousBuildCache.getSources().get(file)
+        ))
+        .collect(Collectors.toUnmodifiableSet());
+  }
+
+  private Optional<SerializedIncrementalCache> readIncrementalCache(Path path) throws IOException {
+    log.debug("Reading incremental cache in from {}", path);
+
+    try (var inputStream = Files.newInputStream(path)) {
+      return Optional.of(objectMapper.readValue(inputStream, SerializedIncrementalCache.class));
+    } catch (NoSuchFileException ex) {
+      log.debug("No file found at {}", path);
+      return Optional.empty();
+    }
+  }
+
+  private void writeIncrementalCache(
+      Path path,
+      SerializedIncrementalCache cache
+  ) throws IOException {
+    log.debug("Writing incremental cache out to {}", path);
+
+    try (var outputStream = Files.newOutputStream(path)) {
+      objectMapper.writeValue(outputStream, cache);
+    }
+  }
+
+  private SerializedIncrementalCache buildIncrementalCache(ProjectInputListing listing) {
+    var dependencyDigests = listing.getDependencySources().stream()
+        .map(SourceListing::getSourceProtoFiles)
+        .flatMap(Collection::stream)
+        .map(this::createSerializedFileDigestAsync)
+        .collect(concurrentExecutor.awaiting())
+        .stream()
+        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+
+    var sourceDigests = listing.getCompilableSources().stream()
+        .map(SourceListing::getSourceProtoFiles)
+        .flatMap(Collection::stream)
+        .map(this::createSerializedFileDigestAsync)
+        .collect(concurrentExecutor.awaiting())
+        .stream()
+        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+
+    return ImmutableSerializedIncrementalCache.builder()
+        .generatedAt(OffsetDateTime.now())
+        .dependencies(dependencyDigests)
+        .sources(sourceDigests)
+        .build();
+  }
+
+  private FutureTask<Entry<Path, String>> createSerializedFileDigestAsync(Path file) {
+    return concurrentExecutor.submit(() -> {
+      log.trace("Generating digest for {}", file);
+      try (var inputStream = Files.newInputStream(file)) {
+        var digest = Digests.sha512ForStream(inputStream);
+        return new SimpleImmutableEntry<>(file, digest);
+      }
+    });
+  }
+
+  private Path getIncrementalCacheRoot() {
+    return temporarySpace.createTemporarySpace("incremental");
+  }
+
+  private Path getPreviousIncrementalCachePath() {
```

Comment:
- [ ] Add version info to path here to prevent crashes if we make further breaking changes to this schema in the future.